bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -349,11 +359,41 @@ impl<T: AsyncRead + AsyncWrite + Unpin + Send> SourceProtocol<T> {
             vmm_ram_range
         );
 
-        // Determine whether we can offer only dirty pages, or if we must offer
-        // all pages.
+        // In the pre-pause phase, it is safe to offer only dirty pages if (1)
+        // there is some prospect of being able to restore the kernel dirty page
+        // bitmap if migration fails, and (2) a prior attempt to restore the
+        // bitmap hasn't failed (thereby rendering the bitmap's contents
+        // untrustworthy). The first prong was checked when the protocol
+        // started, but the second prong requires input from the VM state
+        // driver. If this routine is being called from the pre-pause phase, and
+        // the dirty page map looks viable, ask the state driver if it's OK to
+        // proceed with transmitting only dirty pages.
         //
         // Refer to the giant comment on `RamOfferDiscipline` above for more
         // details about this determination.
+        if *phase == MigratePhase::RamPushPrePause && self.dirt.is_some() {
+            // The state driver should keep the command channels alive until the
+            // migration task exits, so these sends and receives should always
+            // work.
+            self.command_tx
+                .send(MigrateSourceCommand::QueryRedirtyingFailed)
+                .await
+                .unwrap();
+
+            let response = self.response_rx.recv().await.unwrap();
+            match response {
+                MigrateSourceResponse::RedirtyingFailed(has_failed) => {
+                    if has_failed {
+                        self.dirt = None;
+                    }
+                }
```

Comment:
I'm not sure if I love the idea of doing a somewhat complex message-passing handshake thingy to send a bit of state that's only relevant to the code within this module out of this module, into some other task, and then ask it to please give us that bit of state that only we care about back. I don't see any obvious _bugs_ with this,[^1] it just feels quite convoluted and it makes me wonder if there's a simpler solution.

(my [original approach](https://github.com/oxidecomputer/propolis/pull/682/files/721a0e4313a5975065d107040f958b5dfb4864dc#r1569141450) was, for the record, just to stick a `static AtomicBool` in this module and move on with our lives...)

[^1]: It _does_ seem a lot less efficient to stick this in a queue, wait until every other request in the queue ahead of it has been processed, and then wait to get our response back, but the performance effects probably aren't actually meaningful in a protocol where we have to _actually send all the guest memory_... :)

Reply:
It seems very doable just to execute migration protocols on the driver task itself and just omit the message passing entirely. I'm going to try to put that together tomorrow and put up a second PR stacked on this one (assuming it pans out).

Reply:
I've put up a stacked PR, #720, that runs live migration work on the state driver (and makes things a bit more generic so that ~~if~~ when we have multiple protocol versions in the future, the driver doesn't have to care about which one was negotiated). That change gets rid of all the message passing, and should also take care of the other comments above (much less type/unwrap gore).